### DISCUSSION TITLE
Support media type versioning for request/response models

### DISCUSSION THREAD

**Discussion by gmanny** [fastapi/fastapi#8631] [Questions]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
# see "Wanted Code" section
```

### Description

Media type versioning is a recommended practice for versioning a RESTful API across the industry: Zalando [recommends it](https://opensource.zalando.com/restful-api-guidelines/#114), there are [various](https://www.narwhl.com/2015/03/the-ultimate-solution-to-versioning-rest-apis-content-negotiation/) [articles](https://docs.microsoft.com/en-us/archive/msdn-magazine/2014/february/cutting-edge-content-negotiation-and-web-api-for-the-asp-net-mvc-developer) [that](https://www.linkedin.com/pulse/api-versioning-best-practices-muhammed-mokhtar) endorse it. And overall it seems like a very good and flexible solution for versioning APIs:

* It allows versioning individual API endpoints when you need it.
* Even better, you can version input and output schemas independently.
* It allows the rollout of backward-incompatible changes in a backward compatible manner without bending the rules of RESTful API semantics, i.e. without the need to invent a new "object type" to put a new version of the endpoint on a separate URL.
* It can be fully described with an OpenAPI schema, e.g. attached [`openapi.json.zip`](https://github.com/tiangolo/fastapi/files/8283734/openapi.json.zip).

These articles I linked above further explain what I think "Media type versioning" means.

Other good descriptions about what I mean by "media type versioning" can be found here:

* [The Ultimate Solution to Versioning REST APIs: Content Negotiation](https://www.narwhl.com/2015/03/the-ultimate-solution-to-versioning-rest-apis-content-negotiation/)
* [API Design: Media Type Versioning](https://docs.microsoft.com/en-us/azure/architecture/best-practices/api-design#media-type-versioning)

Some of these pages describe only versioning the request or response object, but ideally both should be supported.

### Wanted Solution

* Ability to specify multiple request and response models tied to their respecful content types. (See Wanted Code.)
* Correct generation of the OpenAPI schema based on that. (An example OpenAPI schema is attached to this issue.)
* Parsing of a correct request model based on the request Content-Type.
* A more streamlined access to the "Accept" header than inspecting `Request.headers`
* Automatic setting of the response "Content-Type" header based on the model returned by path operation function.
* The same validation of the returned model as what happens right now for unversioned endpoints.

With a little bit of your direction, I would be thrilled to make a contribution or create a supporting package for this behavior.

If this can be a supporting package, I need a little bit of guidance about how to intercept current flow of schema generation, request parsing and response validation from external code. I can use the code from `InferringRouter` in `fastapi-utils` to learn some of this, but any other high-level pointers would be greatly appreciated.

If you'd prefer contribution, then I would start from an overall discussion on how you see it done, with couple of references to source code.

### Wanted Code

```python
"""
Example implementation of the API versioning using media types.
"""

from fastapi import APIRouter, Body, HTTPException, Request
from pydantic import BaseModel, Field


class InputModelV1(BaseModel):
    "First version of input data"
    name: str = Field(description="Name")
    count: int = Field(description="Count")


class InputModelV2(BaseModel):
    "Second version of input data. Name is split, amount is a float now."
    first_name: str = Field(description="First name")
    last_name: str = Field(description="First name")
    amount: float = Field(description="Total amount")


class OutputModelV1(BaseModel):
    "First version of output data"
    id: int | None = Field(description="Id")
    owner: int | None = Field(description="Owner Id")


class OutputModelV2(BaseModel):
    "Secnod version of output data. Ids moved to strings, owner name supplied as well."
    id: str | None = Field(description="Id")
    owner_id: str | None = Field(description="Owner Id")
    owner_name: str | None = Field(description="Owner Name")


router = APIRouter()


@router.post(
    "/test",
    responses={
        200: {
            "content": {
                "application/json;my_output_model_v=1": {"model": OutputModelV1},
                "application/json;my_output_model_v=2": {"model": OutputModelV2},
            },
            "description": "Return versioned JSON based on either an Accept header or"
            "the version of the input model",
        }
    },
)
def versioned_endpoint_v1(
    req: Request,
    v1: InputModelV1
    | None = Body(None, media_type="application/json;my_input_model_v=1"),
    v2: InputModelV2
    | None = Body(None, media_type="application/json;my_input_model_v=2"),
) -> OutputModelV1 | OutputModelV2 | None:
    "Example endpoint versioned with media types"
    
    # this if block would ideally happen under the hood
    if v1:
        result = do_op_v1(v1)
    elif v2:
        result = do_op_v2(v2)
    else:
        raise HTTPException(status_code=415, detail="Unknown input version")

    # this check would ideally be more abstracted   
    if req.headers.get("accept") == "application/json;my_output_model_v=1":
        return to_v1(result)
    elif req.headers.get("accept") == "application/json;my_output_model_v=2":
        return to_v2(result)
    else:
        # this wouldn't work like this, the output type should be checked before
        # running the operation
        raise HTTPException(status_code=406, detail="Unsupported output version")
```

### Alternatives

Use URL slugs and different path operation functions for versioning.

Other alternatives like versioning based on query parameter or a custom header are not compatible with OpenAPI and would have the same issues with multiple models for request/response body.

### Operating System

Windows

### Operating System Details

This feature is OS-agnostic.

### FastAPI Version

0.75.0

### Python Version

3.10.2

### Additional Context

Issue #3910 describes the "Response" part of what I describe in this issue.

Here's a screenshot of what the Swagger UI looks like for the example [`openapi.json.zip`](https://github.com/tiangolo/fastapi/files/8283734/openapi.json.zip) file I provided:

<img width="986" alt="image" src="https://user-images.githubusercontent.com/1166296/158789677-05f234a0-624f-4fdd-bce2-cdd9c2180e3a.png">

**Comment by gmanny** [comment]
I undrestand that `.zip` files may be suspicious, here's the example OpenAPI schema that I provided as a file:

```json
{
    "openapi": "3.0.2",
    "info": { "title": "Example API", "version": "0.1.1" },
    "paths": {
        "/versioning/test": {
            "post": {
                "tags": ["prototypes"],
                "summary": "Versioned Endpoint V1",
                "description": "Example endpoint versioned with media types",
                "operationId": "versioned_endpoint_v1_versioning_test_post",
                "requestBody": {
                    "content": {
                        "application/json;my_input_model_v=1": {
                            "schema": {
                                "$ref": "#/components/schemas/InputModelV1"
                            }
                        },
                        "application/json;my_input_model_v=2": {
                            "schema": {
                                "$ref": "#/components/schemas/InputModelV2"
                            }
                        }
                    }
                },
                "responses": {
                    "200": {
                        "description": "Return versioned JSON based on either an Accept header orthe version of the input model",
                        "content": {
                            "application/json;my_output_model_v=1": {
                                "schema": {
                                    "$ref": "#/components/schemas/OutputModelV1"
                                }
                            },
                            "application/json;my_output_model_v=2": {
                                "schema": {
                                    "$ref": "#/components/schemas/OutputModelV2"
                                }
                            }
                        }
                    },
                    "422": {
                        "description": "Validation Error",
                        "content": {
                            "application/json": {
                                "schema": {
                                    "$ref": "#/components/schemas/HTTPValidationError"
                                }
                            }
                        }
                    }
                }
            }
        }
    },
    "components": {
        "schemas": {
            "HTTPValidationError": {
                "title": "HTTPValidationError",
                "type": "object",
                "properties": {
                    "detail": {
                        "title": "Detail",
                        "type": "array",
                        "items": {
                            "$ref": "#/components/schemas/ValidationError"
                        }
                    }
                }
            },
            "InputModelV1": {
                "title": "InputModelV1",
                "required": ["name", "count"],
                "type": "object",
                "properties": {
                    "name": {
                        "title": "Name",
                        "type": "string",
                        "description": "Name"
                    },
                    "count": {
                        "title": "Count",
                        "type": "integer",
                        "description": "Count"
                    }
                },
                "description": "First version of input data"
            },
            "InputModelV2": {
                "title": "InputModelV2",
                "required": ["first_name", "last_name", "amount"],
                "type": "object",
                "properties": {
                    "first_name": {
                        "title": "First Name",
                        "type": "string",
                        "description": "First name"
                    },
                    "last_name": {
                        "title": "Last Name",
                        "type": "string",
                        "description": "First name"
                    },
                    "amount": {
                        "title": "Amount",
                        "type": "number",
                        "description": "Total amount"
                    }
                },
                "description": "Second version of input data. Name is split, amount is a float now."
            },
            "OutputModelV1": {
                "title": "OutputModelV1",
                "type": "object",
                "properties": {
                    "id": {
                        "title": "Id",
                        "type": "integer",
                        "description": "Id"
                    },
                    "owner": {
                        "title": "Owner",
                        "type": "integer",
                        "description": "Owner Id"
                    }
                },
                "description": "First version of output data"
            },
            "OutputModelV2": {
                "title": "OutputModelV2",
                "type": "object",
                "properties": {
                    "id": {
                        "title": "Id",
                        "type": "string",
                        "description": "Id"
                    },
                    "owner_id": {
                        "title": "Owner Id",
                        "type": "string",
                        "description": "Owner Id"
                    },
                    "owner_name": {
                        "title": "Owner Name",
                        "type": "string",
                        "description": "Owner Name"
                    }
                },
                "description": "Secnod version of output data. Ids moved to strings, owner name supplied as well."
            },
            "ValidationError": {
                "title": "ValidationError",
                "required": ["loc", "msg", "type"],
                "type": "object",
                "properties": {
                    "loc": {
                        "title": "Location",
                        "type": "array",
                        "items": { "type": "string" }
                    },
                    "msg": { "title": "Message", "type": "string" },
                    "type": { "title": "Error Type", "type": "string" }
                }
            }
        }
    }
}
```

**Comment by plankthom** [comment]
I contributed PR #4727 that merges open-api descriptions for routes that overlap on path and method.
In the current version only the last operation description for each path and method is kept in the open-api spec.

This is a solution for the open api spec generation, while custom router classes can solve the *routing on Accept header* part of this issue.

**Comment by skrech** [comment]
Hello, I was about to open a new Issue with the same request. I'm happy that other people are also seeking a solution for how to define model-per-media-type in `responses`. 
Currently, FastAPI assumes that responses (and also requests) for a given status_code are always from a single media-type. However, in a lot of cases this doesn't reflect the reality. Even more, OpenAPI has great support for defining multiple media type with different schema per status-code, so it'd be great for FastAPI to support that in more streamlined way.

Although content-negotiation is closely tied to responses' media type, I don't perceive this as a big issue, because it's relatively easy to create content negotiation logic with dependencies and Accept header. However, documenting the various models per media-type in OpenAPI spec is almost impossible (and especially if we want to structure the resulting document with $ref and components), because there aren't any such "hooks" provided by FastAPI. 

I think a great first step towards better media type support would be to support "model" key in `responses/content/media-type` as in the suggested variant from @gmanny:
```
responses={
        200: {
            "content": {
                "application/json;my_output_model_v=1": {"model": OutputModelV1},
```

PS -- I've also checked the MR from @plankthom but the suggested solution involves, IMO, a lot of path operation metadata duplication (descriptions, summaries, examples, etc) and it seems a bit error-prone, especially with various parameters defined in different functions while accumulated in a single openapi description. Additionally, it seems that merging is possible only if using `add_api_route` directly, won't be working with the standard `get`, `put`, etc decorators.
However, the idea to dispatch to functions based on expected "main" response seems nice!

**Reply by g0di** [reply]
`documenting the various models per media-type in OpenAPI spec is almost impossible`

Confirmed =>

```python
from typing import Any

from fastapi import FastAPI
from pydantic import BaseModel, TypeAdapter


class MyModel(BaseModel):
    foo: str
    bar: str


app = FastAPI()


@app.get(
    "/helloworld",
    responses={
        200: {
            "content": {
                "application/v2+json": {
                    "schema": TypeAdapter(list[MyModel]).json_schema()
                }
            }
        }
    },
)
def say_hello() -> Any:  # noqa: ANN401
    return "Hello!"
``` 

Is showing this
![image](https://github.com/tiangolo/fastapi/assets/2545380/89799303-8a09-4a5a-a95d-b69b7092cb87)
>The `example` section displayed is also wrong, it shows `[ string ]` where `string` should actually be the `MyModel` schema

It tried to see quickly in the code how models are actually handled to try to replicate but I gave up.

**Comment by zevisert** [comment]
I'm looking for something like this as well, IIUC this feature request is for a router to use multiple dispatch on routes that use the same path and method, right?

**Comment by allartk** [comment]
For anyone ending here from a search engine looking for a way to document different media types (eg application/json and application/ld+json) , the example in https://github.com/tiangolo/fastapi/discussions/7070#discussioncomment-5134040 might help you.

For me, as a new user, the documentation in https://fastapi.tiangolo.com/advanced/additional-responses/#combining-information comes close, but the gotcha for me was the presence of the schema method.

**Comment by midprasanta** [comment]
Here the problem is openapi spec generates beutifully but api does not work. When you invoke the API , pydantic gives error "Input should be a valid dictionary or object to extract fields from" 
So If you specify this content type(application/json;my_input_model_v=1) in postman request header with this custom mediatype, it does not work.

**Comment by exhuma** [comment]
It is great to see interest in this. I have been using versioned media-types with content negotiation for well over ten years now with incredible success. It has made the API extremely stable and has enabled us to make significant changes without breaking backwards incompatibility in many occasions.

I am currently fiddling around with a solution for FastAPI but have only started. My strategy before stumbling across this post was to write an independent middleware. I also looked into witing it as ASGI middleware to make it available for other frameworks but that prevented me from using many of the FastAPI conveniences so I am experimenting again with FastAPI directly.

In any case, I can bring a bit of experience to this change.

While I really like the proposed example of @gmanny  I would prefer to find a solution that can optionally dispatch incoming requests depending on the result of content-negotiation. While this can still be done in application-code, it would be nice to have a standardised way of handling many media-types. We have a couple of functions that handle many media-types that have accumulated over time. And currently the route-implementations look really ugly with a bunch of `if/else-if` statements. Even the new `match` statement only improves this marginally.

**Comment by kimlehtinen** [comment]
Any updates on this? I'm looking for this exact thing by following [Zalando API guidelines](https://opensource.zalando.com/restful-api-guidelines/#114) as @gmanny described.
